Faster+safer use of unsafe package + non-unsafe fallback implementations

daboyuka opened the pull request:
Pardon the unhelpful PR title; wasn't sure what to call this...

Removed the `unsafeBytesToString` function, instead encapsulating `unsafe` functionality in a separate set of files (`fastbytes*`), thus preventing misuse by client code. Included are fast implementations of `BytesEqualStr`, which takes one string and one []byte argument, and `BytesParseFloat` and `BytesParseInt`, which each take a `[]byte` argument. None of these functions let unsafe strings escape, thus improving overall safety.

The main benefit of this refactoring is the ability to have non-unsafe fallback versions of these functions for platforms without the `unsafe` package (such as Google App Engine). Such implementations are included in `fastbytessafe.go`.

A somewhat faster implementation of BytesEqualStr is included than was previously used in the codebase. Sadly, string equality checks are a small portion of overall `Get*()` cost now, so there is little net performance impact. A significantly faster implementation of BytesParseInt is also included, but has little net impact for the same reason.

No noticeable net change in the benchmark, however the code is now cleaner, safer, and more portable.

Note that this patch exports `Bytes{EqualStr,ParseInt,ParseFloat}`, as they may be of use to client code. This is far safer than the old method of exporting `GetUnsafeString`, as these new functions are safe.

**Benchmark before change**:
```
BenchmarkJsonParserLarge-8 	  100000	     62860 ns/op	       0 B/op	       0 allocs/op
BenchmarkJsonParserMedium-8	  500000	     10639 ns/op	       0 B/op	       0 allocs/op
BenchmarkJsonParserSmall-8 	10000000	       985 ns/op	       0 B/op	       0 allocs/op
```

**Benchmark after change**:
```
BenchmarkJsonParserLarge-8 	  100000	     62428 ns/op	       0 B/op	       0 allocs/op
BenchmarkJsonParserMedium-8	 1000000	     10779 ns/op	       0 B/op	       0 allocs/op
BenchmarkJsonParserSmall-8 	10000000	       996 ns/op	       0 B/op	       0 allocs/op
```

daboyuka commented:
Note: this PR is not ready yet; I just wanted you to be aware of what I was working on. The actual PR consists of only the last two commits (the rest are my earlier PR commits, since this PR is dependent on some of them).

daboyuka commented:
Oh yeah, and the "Simpler (faster?) true/false/null literal checking" one I haven't made a separate PR for yet.

daboyuka commented:
Note: there may be some issue merging this with PR #28 at this line:

`value := unsafeBytesToString(data[offset : endOffset+end])` (line 226 on this branch)

The equivalent line in PR #28 should be selected in the merge.

daboyuka commented:
Looks like there's also a conflict on line 124; use this version of the line: `strbytes.Equal(keys[level-1], data[keyBegin:keyEnd]) {`

daboyuka commented:
Note: this PR also removes `GetUnsafeString` because of the restructuring of unsafe functions. It can be added back if that's the final decision, though I would at least recommend it go in the new `strbytes` package with a non-unsafe version for platforms without the `unsafe` package, based on build flags.

daboyuka commented:
Any thoughts on this PR? Is there something about it you don't like or that should be changed? I don't mean to rush you at all, just curious.

buger commented:
Yeah, change is big, did not had time to write comment till now. 

So, main idea here is to provide compatibility with environments that do not support `unsafe` package, right? And i like this idea. 
What i do not like is having it as a separate package, it make contributing way harder (you need to rewrite paths to package during merge every time, and this is kinda additional dependency, even internal), i prefer to include those packages directly to `jsonparser` namespace. 
Also "moreUnsafe" sounds dangerous :) Do you know real difference between `unsafe` and `moreUnsafe` versions? If you can give me link it will be perfect. 

Note that this PR will be merged manually, so probably you do not need to make more changes here, just answer questions above.

Also when you provide benchmarks, do not forget add `-test.benchmem` option to show memory allocation, some `unsafe` usage was made to remove allocations, its not only about CPU. 

And as always thank you so much for all your contributions!

daboyuka commented:
Sure, these are good questions, I'm happy to try to answer. First, though, I have one request: would you mind to let me update my PR before you merge so that it won't require a manual merge? I know this code well, so it's easier and safer for me to do the update. Once you're satisfied with the plan, I can make the changes and notify you here.

Compatibility with other environments is one main point. The other is to encapsulate the `unsafe` functionality as much as possible to limit the chances for bad bugs.

Moving everything into the main package seems fine to me. The only reason I did it that way was to enforce encapsulation, but there are other ways to do that (including just not using the internal functions). I can move it back out.

`moreUnsafe` is based on this StackOverflow post: http://stackoverflow.com/a/31484416. I've included an explanation of how it in comparison to `unsafe` below.

Thanks for the info about `-test.benchmem`, I didn't know that one. I'll be sure to use it in the future.

`moreUnsafe` explanation:
As you know, both slices and strings are stored as a small headers that each contain a pointer to the actual data. The layout of a slice header is (currently) `(pointer) (len) (cap)`, whereas the layout of a string header is `(pointer) (len)` (both `(pointer)` and (`len`) are of the same type and alignment in both cases).

The `unsafe` method operates by constructing a new string header using the same pointer as a slice header; this is quite cheap, as it only requires copying two values (pointer and length) to a new stack-allocated header. This utilizes `reflect` types `SliceHeader` and `StringHeader` to be sure everything is copied right, and so it will work no matter how the pointer/len/cap fields are arranged within the headers.

The `moreUnsafe` method operates on the assumption that (pointer) and (len) in both kinds of header 1) have the same types, 2) have the same alignments, and 3) appear in the same order (all of which are true since the beginning of Go until now, I think). Given these facts, `moreUnsafe` simply reinterprets the slice's header as a string header in-place: the (pointer) and (len) fields are where they are supposed to be, and the (cap) field is just ignored. This approach is slightly faster because it requires no operations of any kind (it's just a pointer reinterpretation). For short strings, this can be noticeably faster than `unsafe` (see the benchmark I added in that package).

buger commented:
Can you imagine situation when moreUnsafe version breaks? For example length of string that contains unicode characters can differ from length of same []byte, can you make some unicode tests?

I'm ok if you will make this changes, but not that currently there is conflicts with current master, so maybe make sense to create new PR, or solve them first, here. Pls move all the functions to `jsonparser` namespace, and ensure that your changes does not add additional memory allocations. 

Thanks!

daboyuka commented:
The length of a string that contains Unicode _cannot_ differ from the length of the same []byte. For example: http://play.golang.org/p/5lgSA2ko8-. The only way `moreUnsafe` would break is if Go changed the slice/string header structs in C, which hasn't happened yet, and could only happen in a new version of Go. Despite being called `moreUnsafe`, I think it's safe enough :).

I'll work on the merge soon, maybe later today (Friday). Thanks!

buger commented:
I see, thanks! I thought len(string) will print actual length for runes... You have to use `utf8.RuneCountInString` to do it.

http://play.golang.org/p/ckqYkyJFCQ

daboyuka commented:
For what it's worth, I've been paying attention to that distinction as I read the code, and so far I haven't seen any UTF-related bugs. That proves nothing, though :).

daboyuka commented:
I reworked the branch to address your suggestions. You may find the updated benchmarks in the original post.

* Updated to the latest version of buger/jsonparser/master (no more merge conflicts)
* Removed the `strbytes` package and moved the files into the main `jsonparser` package
* Simplified the source code a good bit
* Developed a custom implementation of ParseInt that operates directly on a []byte and is significantly faster than the `strconv` even when ignoring string conversion (it gains speedup by hard-coding base 10 and 64-bit integer return).

daboyuka commented:
Oops, don't merge yet, I want to add one more thing.

daboyuka commented:
Ok, my changes in this PR are finally finished. Let me know if there are any further issues.

daboyuka commented:
Fixed some merge conflicts with respect to latest master, so this will merge again.

daboyuka commented:
Hi @buger, is there anything missing from this PR that you'd like me to work on yet?

daboyuka commented:
Note: if you merge #40 first, a block of code in `searchKeys` will conflict; here's what the conflict fix should look like:

```
if data[i] == ':' {
	key := data[keyBegin:keyEnd]

	// for unescape: if there are no escape sequences, this is cheap; if there are, it is a
	// bit more expensive, but causes no allocations unless len(key) > unescapeStackBufSize
	if keyUnesc, err := Unescape(key, stackbuf[:]); err != nil {
		return -1
	} else {
		if keyLevel == level-1 && // If key nesting level match current object nested level
			BytesEqualStr(&keyUnesc, keys[level-1]) {
			keyLevel++
			// If we found all keys in path
			if keyLevel == lk {
				return i + 1
			}
		}
	}
} else {
	i--
}
```

buger commented:
Merged with some refactoring, thanks!
